src/libcore/ffi.rs

```diff
@@ -40,3 +43,182 @@ impl fmt::Debug for c_void {
+// but also ensure it cannot be used elsewhere, the trait needs to be public
+// within a private module. Once RFC 2145 has been implemented look into
+// improving this.
+mod sealed_trait {
+    /// Trait which whitelists the allowed types to be used with [VaList::arg]
+    ///
+    /// [VaList::va_arg]: struct.VaList.html#method.arg
+    #[unstable(feature = "c_variadic",
+               reason = "the `c_variadic` feature has not been properly tested on \
+                         all supported platforms",
+               issue = "27745")]
+    pub trait VaArgSafe {}
+}
+
+macro_rules! impl_va_arg_safe {
+    ($($t:ty),+) => {
+        $(
+            #[unstable(feature = "c_variadic",
+                       reason = "the `c_variadic` feature has not been properly tested on \
+                                 all supported platforms",
+                       issue = "27745")]
+            impl sealed_trait::VaArgSafe for $t {}
+        )+
+    }
+}
+
+impl_va_arg_safe!{i8, i16, i32, i64, usize}
+impl_va_arg_safe!{u8, u16, u32, u64, isize}
+impl_va_arg_safe!{f64}
+
+#[unstable(feature = "c_variadic",
+           reason = "the `c_variadic` feature has not been properly tested on \
+                     all supported platforms",
+           issue = "27745")]
+impl<T> sealed_trait::VaArgSafe for *mut T {}
+#[unstable(feature = "c_variadic",
+           reason = "the `c_variadic` feature has not been properly tested on \
+                     all supported platforms",
+           issue = "27745")]
+impl<T> sealed_trait::VaArgSafe for *const T {}
+
+#[cfg(not(stage0))]
+impl<'a> VaList<'a> {
+    /// Advance to the next arg.
+    #[unstable(feature = "c_variadic",
+               reason = "the `c_variadic` feature has not been properly tested on \
+                         all supported platforms",
+               issue = "27745")]
+    pub unsafe fn arg<T: sealed_trait::VaArgSafe>(&mut self) -> T {
+        va_arg(self)
+    }
+
+    /// Copy the `va_list` at the current location.
+    #[unstable(feature = "c_variadic",
+               reason = "the `c_variadic` feature has not been properly tested on \
+                         all supported platforms",
+               issue = "27745")]
+    pub unsafe fn copy<F, R>(&mut self, f: F) -> R
+            where F: for<'copy> FnOnce(VaList<'copy>) -> R {
+        #[cfg(any(all(not(target_arch = "aarch64"), not(target_arch = "powerpc"),
+                      not(target_arch = "x86_64")),
+                  windows))]
+        let mut ap = MaybeUninit::uninitialized().into_inner();
+        #[cfg(all(any(target_arch = "aarch64", target_arch = "powerpc", target_arch = "x86_64"),
+                  not(windows)))]
+        let mut ap_inner: VaListImpl = MaybeUninit::uninitialized().into_inner();
+        #[cfg(all(any(target_arch = "aarch64", target_arch = "powerpc", target_arch = "x86_64"),
+                  not(windows)))]
+        let mut ap = VaList(&mut ap_inner);
```

Comment:
This should be using `.get_mut()` on a `MaybeUninit`.

Reply:
Actually it should be `as_mut_ptr`, right?

Reply:
Oh you actually have a reference there. Why do you want a reference to uninitialized data? So far this should be considered UB, until we have made the decision to allow it.

Reply:
Hmpf. Maybe we can `#[derive(Default)]` on the various definitions of `VaListImpl`?

Reply:
Or use a raw pointer in `VaList`.

I have no idea what this is all about, just making some guesses here.^^

Reply:
It's semantically a reference, with a lifetime. Really, `va_copy` should be returning the copy instead of taking an output pointer!

Reply:
Oh, you want `&out` or so. I guess you could emulate it with
```
struct VaList<'a>(*mut VaListImpl, PhantomData<&'a mut VaListImpl>)
```

Reply:
No, only `va_copy` does, *because* one of its arguments is an out pointer. All other uses want `&mut`. This is why I said I prefer `va_copy` returning the copy.

Reply:
Note that I still prefer making `va_copy` return the copy, instead of having the Rust code pass a mutable reference into it.